src/c-writer.cc

```diff
@@ -170,17 +170,63 @@ int GetShiftMask(Type type) {
   // clang-format on
 }
 
+/* This function is the default behavior for name_to_index_. For single .c
+ * output, this function returns a vector filled with 0. For multiple .c
+ * outputs, this function sorts all non-imported functions in the module by
+ * their names, and then divides all non-imported functions into equal-sized
+ * buckets (# of non-imported functions / # of .c outputs) based on the sorting.
+ */
+static std::vector<size_t> default_name_to_index(
+    const std::vector<Func*>& funcs,
+    const size_t num_imports,
+    const size_t num_streams) {
+  std::vector<size_t> result;
+  result.resize(funcs.size());
+  if (num_streams == 1) {
+    return result;
+  }
+
+  std::map<std::string, Index> sorted_functions;
+  size_t non_imported_funcs = funcs.size() - num_imports;
+  size_t bucket_size = non_imported_funcs / num_streams +
+                       (non_imported_funcs % num_streams ? 1 : 0);
+  Index func_index = 0;
+  for (const Func* func : funcs) {
+    sorted_functions.insert({func->name, func_index});
+    ++func_index;
+  }
+  Index sorted_func_index = 0;
+  for (const auto& [func_name, index] : sorted_functions) {
+    bool is_import = index < num_imports;
+    if (!is_import) {
+      result.at(index) = sorted_func_index / bucket_size;
+      ++sorted_func_index;
+    }
+  }
+  return result;
+}
+
 class CWriter {
  public:
-  CWriter(Stream* c_stream,
+  CWriter(std::vector<Stream*>&& c_streams,
           Stream* h_stream,
+          Stream* h_impl_stream,
           const char* header_name,
+          const char* header_impl_name,
           const WriteCOptions& options)
       : options_(options),
-        c_stream_(c_stream),
+        c_streams_(std::move(c_streams)),
         h_stream_(h_stream),
-        header_name_(header_name) {
+        h_impl_stream_(h_impl_stream),
+        header_name_(header_name),
+        header_impl_name_(header_impl_name) {
     module_prefix_ = MangleModuleName(options_.module_name);
+    if (c_streams_.size() != 1 && options.name_to_index) {
+      name_to_index_ = options.name_to_index;
+    } else {
+      name_to_index_ = std::bind(&default_name_to_index, std::placeholders::_1,
```

Comment:
Now that `default_name_to_index` is an ordinary function, I think we can get rid of the call to `std::bind` and just say `name_to_index_ = default_name_to_index;`